test/images/README.md

```diff
@@ -0,0 +1,129 @@
+# Building Kubernetes test images
+
+## Overview
+
+All the images found here are used in Kubernetes tests that ensures its features and functionality.
+The images are built and published as manifest lists, allowing multiarch and cross platoform support.
+
+## Prerequisites
+
+In order to build the docker test images, a Linux node is required. The node will require `make`
+and `docker (version 18.06.0 or newer)`. Manifest lists were introduced in 18.03.0, but 18.06.0
+is recommended in order to avoid certain issues.
+
+The node must be able to push the images to the desired container registry. Depending on the
+container registry, it might require a different authentication method. For dockerhub, this
+can be done by running the command:
+
+```bash
+    docker login -u your-awesome-username -p anAwesomerPassword
+```
+
+Windows Container images are not built by default, since they cannot be built on Linux. For
+that, a Windows node with Docker installed and configured for remote management is required.
+
+### Windows node setup
+
+In order to build the Windows container images, a node with Windows 10 or Windows Server 2019
+with the latest updates installed is required. The node will have to have Docker installed,
+preferably version 18.06.0 or newer.
+
+Note which version of Docker is installed on the Windows node: Enterprise Edition, or
+Community Edition. If it is EE, the container OS version and the host's OS version **must** match
+in order to start the containers. Currently, the Windows Containers are built based on the
+`mcr.microsoft.com/windows/servercore:ltsc2019` image.
```

Comment:
Don't we also want to build images for newer Windows OS images from the semi-annual release channels?

Reply:
That's going to be more difficult,

Currently, we officially support  WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.

Now, adding multiple OS versions to the image building process was something I've had in mind, but it's not quite as simple. The difficult part would be Windows node(s). Currently, there would be 2 ways this could be done:

- have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images). This would also be a bit more difficult to implement and use, you would have to somehow map hosts with what image versions to build.
- have a single Windows node with docker CE installed. This is what I have been using, I had a Windows 10 node for building images for both 1803 and 1809 and creating manifest lists containing both of them. Now, the disadvantage to this, would be the build time. It takes a lot longer to build an image using this method.

At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.

One thing we could probably do and if there's a need for it, would be to parameterize the base Windows container image, so, if anyone wants to build on top of another Windows OS version, they can just set that parameter in the make command.

Reply:
> Currently, we officially support WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.

Hmm..my concern is that if sig-windows supports a new version of Windows image in the next release, this would still need to be updated to build and test both SAC and LTS images. Is my understanding incorrect?

> * have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images).

This may not be that big of problem if we build the images using GCE VMs @pjh 



> At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.

I'm not familiar with the constraints (or lack thereof) docker CE. How does Docker CE handle building images for multiple OS versions? Does that also mean it can start containers even though versions don't match (without having any issues)?

Reply:
> 
> 
> > Currently, we officially support WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.
> 
> Hmm..my concern is that if sig-windows supports a new version of Windows image in the next release, this would still need to be updated to build and test both SAC and LTS images. Is my understanding incorrect?

Right, it would depend on sig-windows to decide what versions are officially supported. But that would not only imply that we'll have to build images for both versions, but we would also have to add CI jobs for those versions as well.

> 
> > * have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images).
> 
> This may not be that big of problem if we build the images using GCE VMs @pjh
> 
> > At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.
> 
> I'm not familiar with the constraints (or lack thereof) docker CE. How does Docker CE handle building images for multiple OS versions? Does that also mean it can start containers even though versions don't match (without having any issues)?

Yeah, you can start containers even though versions don't match, within some boundaries. At one point, I had to deliberately install the Windows updates on my Windows 10 node in order to be able to build the latest WS 2019 images.

Reply:
Maybe I missed something, but I don't see how sig-windows could escape supporting multiple versions (or supporting both SAC and LTSC). Do we not expect to build/integrate features in released only in SAC at all...? @PatrickLang 

Hm....so Docker CE has looser constraints, but is there guarantee that the containers (with mismatched versions) would run without any issue? Trying to understand where the difference comes from. Is there any documentation that I could read?

Reply:
Ok, so, I have a bit more information about this.

First of all, the reason why multiple image versions were supported on my Windows 10 node is because Hyper-V isolation is enabled by default on Windows client, while Windows Server uses process isolation by default. See the explanation here: https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file (at the "Options for the runtime" section).

Apparently, I can do this on my Windows Server 2019 node without any issue:

    docker run --isolation=hyperv -d --name test-1803 --entrypoint ping microsoft/windowsservercore:1803 -t localhost
    db99b69779941dd55e9a2877817e1597ab845c87ea477873f0230d4e61f01938
    docker run --isolation=hyperv -d --name test-2019 --entrypoint ping mcr.microsoft.com/windows/servercore:ltsc2019 -t localhost
    c57b09fdbafb9c932a34d6729c6ec3a6bb2064c4c300acb50ff820900b0d58c5

Interestingly, I cannot do the same for 1903 images:

    docker run --isolation=hyperv -d --name test-2019 --entrypoint ping mcr.microsoft.com/windows/servercore:1903-amd64 -t localhost
    Unable to find image 'mcr.microsoft.com/windows/servercore:1903-amd64' locally 1903-amd64: Pulling from windows/servercore
    C:\Program Files\Docker\docker.exe: a Windows version 10.0.18362-based image is incompatible with a 10.0.17763 host.
    See 'C:\Program Files\Docker\docker.exe run --help'.

So, it seems that for newer images, a newer Windows Server node is still required.

https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container

Reply:
@yujuhong I've added a commit to this PR (https://github.com/kubernetes/kubernetes/pull/76838/commits/0c744f5d2054605c9c6af38a31815de1b11405eb) which adds support for building for multiple Windows OS versions. I've tried it out for the ``busybox`` version, for WS 1803 and 1809 (since I have a WS 1809 node). Here's the resulting manifest list: https://paste.ubuntu.com/p/q7dhswZcdc/

I've also updated the Windows section in the README to reflect these things.

Is there anything else we need for this PR?

Reply:
Windows Server 1803 - no support. We're not testing that in Kubernetes and declared Windows Server 2019 stable.

Windows Server 2019 & Windows Server version 1809 - same thing, different branding. The containers are interchangeable.

Windows Server version 1903 - let's discuss in SIG. There's no testing for it yet, but it's feasible

Reply:
Ack. I think it'd be valuable to support SAC releases (1903, etc.) to support new features. I agree this should be determined by sig-windows, and was hoping that the doc can be more generic so that we could augment supported versions when needed.